doc/compiler/aot/Diagnostics.md

```diff
@@ -0,0 +1,396 @@
+`00007F5682801364` is the instance of the `TR_RelocationRecord` type
+(or subtype); `00007F56A8BB68FE` is the address in the code the
+relocation is applied to. Not all relocations involve updating the
+code; some will use the relocated address to do other tasks such as
+registering Runtime Assumptions.
+
+# Common reasons for AOT bugs
+
+### Missing or incorrect relocation
+When adding a new feature, or enabling a disabled code path, it is
+important to think about AOT. If the compiler generates or emits
+something that is **not** position **independent**, then the compiled
+code may not work in another JVM instance. Depending on the position
+dependent code, the error may manifest itself in a way that is not
+trivial to understand. An example of this might be if the compiler
+materializes a `J9Class` pointer using some means other than the
+existing front end queries that do ensure that a relocation record
+is generated, and emits it into the code. Even when the method is
+loaded from the SCC in the current instance some time later, there
+won't be a problem because the class pointer will be valid. However,
+in another JVM instance, that same class may be located at a
+different address.
+
+### Missing or incorrect Validation
+In a similar vein, if the compiler generates or emits something based
+on the environment in one JVM instance, that assumption may not hold
+in another JVM instance. Bugs due to missing validations are notoriously
+difficult to pin down. An example of this might be if the compiler
+makes an assumption about the relationship between two classes.
+Because of the dynamic nature of Java, that relationship is not
+guaranteed to hold in another JVM instance. As before, when the method is
+loaded in the current instance there won't be a problem.
+
+### Invalid code
+This can happen when one tries to enable code that was disabled under
+AOT. Because of the requirement that all interactions between the
+compiler and the runtime have to be recorded in some manner, AOT
+compilations often have assumptions that lean towards being more
+conservative. When trying to allow certain optimizations, if not all
+areas are updated to have a consistent set of assumptions, it can lead
+to problems. Often, bugs of this nature are easier to deal with as
+they will also occur in the compile run.
+
+# Debugging Strategies
+Much of the strategies one uses to narrow down the problematic method in
+a JIT compile apply here. However, there are some additional options
+to know about
+- `limit=`, `exclude=`, `limitFile=` apply only to compilations. It is
+worth noting that this will limit both JIT **and** AOT compilations
+regardless of whether it's set on `-Xjit` or `-Xaot`.
+- `loadLimit=`, `loadExclude=`, `loadLimitFile=` apply only to loads;
+these can only be set on `-Xaot`.
+
+Once the problem method is found, one should try and see if the problem
+narrows down to one the three common problems above.
```

Comment:
_one of the three common problems above_?

Reply:
Update in force push.